### PR TITLE
table: support checksum modes

### PR DESCRIPTION
**Pull request by skyzh**
Will be ready for review when table implementation has been merged.

I plan to support all badger checksum options `ChecksumVerificationMode` in this PR

```go
const (
	// NoVerification indicates DB should not verify checksum for SSTable blocks.
	NoVerification ChecksumVerificationMode = iota
	// OnTableRead indicates checksum should be verified while opening SSTtable.
	OnTableRead
	// OnBlockRead indicates checksum should be verified on every SSTable block read.
	OnBlockRead
	// OnTableAndBlockRead indicates checksum should be verified
	// on SSTable opening and on every block read.
	OnTableAndBlockRead
)
```

Signed-off-by: Alex Chi <iskyzh@gmail.com>

### CODE REVIEW

**Comment by skyzh** [comment]
@Connor1996 @BusyJay PTAL, thanks!

File: src/table.rs [hunk 1]
```diff
@@ -256,16 +271,24 @@ impl TableInner {
...
-            // The checksum is calculated for actual data + entry index + index length
-            data: data.slice(..read_pos + 4),
+            data,
             entry_offsets,
             checksum_len,
             checksum,
-        }))
+        });
+
+        if matches!(self.opts.checksum_mode, OnTableAndBlockRead | OnTableRead) {
```
Comment (Connor1996) [inline src/table.rs:287]:
```suggestion
        if matches!(self.opts.checksum_mode, OnTableAndBlockRead | OnBlockRead) {
```

**Review by Connor1996 — APPROVED** [review]
LGTM

File: src/table.rs [hunk 2]
```diff
@@ -126,19 +128,25 @@ impl TableInner {
             index: TableIndex::default(),
             index_start: 0,
             index_len: 0,
-            opts,
+            opts: opts.clone(),
         };
         inner.init_biggest_and_smallest()?;
-        // TODO: verify checksum
+
+        if matches!(opts.checksum_mode, OnTableAndBlockRead | OnTableRead) {
```
Comment (BusyJay) [inline src/table.rs:135]: Why not `inner.opts.checksum_mode`? So you don't need to clone it at L131.

File: src/lib.rs [hunk 3]
```diff
@@ -7,7 +7,7 @@ mod format;
 mod levels;
 mod memtable;
 mod ops;
-mod opt;
+pub mod opt;
```
Comment (BusyJay) [inline src/lib.rs:10]: How about also reexporting it? So users don't need to write two lines for importing from agatedb.

File: src/table/tests.rs [hunk 4]
```diff
@@ -344,3 +356,17 @@ fn test_table_big_values() {
     // TODO: support max_version in table
     // assert_eq!(n, table.max_version());
 }
+
+#[test]
+fn test_table_checksum() {
+    let mut rng = thread_rng();
+    let opts = get_test_table_options();
```
Comment (BusyJay) [inline src/table/tests.rs:363]: Original implement seems to use `OnTableAndBlockRead`.

**Comment by skyzh** [comment]
@BusyJay PTAL, thanks!

File: src/table.rs [hunk 5]
```diff
@@ -150,6 +158,11 @@ impl TableInner {
             index_len: 0,
         };
         inner.init_biggest_and_smallest()?;
+
+        if matches!(opts.checksum_mode, OnTableAndBlockRead | OnTableRead) {
```
Comment (BusyJay) [inline src/table.rs:162]: You can still use `inner.opts` to get rid of `opts.clone()`.

**Review by BusyJay — COMMENTED** [review]
Rest LGTM

**Review by BusyJay — APPROVED** [review]

**Pull request merged**